### PR TITLE
app:glusterfs: better error propagation from brick allocation

### PR DESCRIPTION
**Pull request by obnoxxx**
Signed-off-by: Michael Adam <obnox@redhat.com>

### CODE REVIEW

**Comment by obnoxxx** [comment]
Not sure if it's correct.
But it may bring some clarity to issues #466  and #331 .

**Comment by obnoxxx** [comment]
ok. patch is buggy... doesn't compile, var out of scope.  closing. will re-PR once I have a better patch ;-)

File: apps/glusterfs/volume_entry.go [hunk 1]
```diff
@@ -304,6 +304,11 @@ func (v *VolumeEntry) Create(db *bolt.DB,
 			break
 		}
 	}
+
+	if err != nil {
+		return err
+	}
+
```
Comment (lpabon) [inline apps/glusterfs/volume_entry.go]: I think the correct fix will probably be to add what you have, but then move the logic of _ErrNoSpace_ to v.allocBricksInCluster(), then change line 300 to say something like: if err == ErrNoSpace, then continue, else get out!

Comment (lpabon) [inline apps/glusterfs/volume_entry.go]: You will also need to make sure unit tests are adjust to expect the correct paths to test the change

**Comment by obnoxxx** [comment]
@lpabon : I have been looking over the code quite some time now. Here are my thoughts:

The v.allocBricksInCluster() function does currently return three different errors that indicate that the current cluster can provide space for the requested volume size:
- ErrNoSpace
- ErrMinimumBrickSize
- ErrMaxBricks
These errors justify continuing with the next cluster as a candidate for the volume.
All other errors are genuine, serious errors, like database errors,  that justify aborting the creation.

So, your proposal is to map the three 'valid' errors to ErrNoSpace in v.allocBricksInCluster(), and then just treat the ErrNoSpace case in the caller (Create, or Expand) ?

It may be worth keeping the different errors possibly returned from allocBricksInCluster(), for the better error propagation. (At least the unit tests for expand use /test that.) In that case, the loop in Create() just needs to check three 'valid' errors instead of one.

On the other hand, if we mapped the valid errors to ErrNoSpace in allocBricksInCluster(), then we could just get rid of ErrMaxBricks and ErrMinumumBrickSize. That would be a simplification, but a loss of granularity in error reporting and possibly debuggability?

Not 100% convinced yet what the better change would be, but I'll propose a patch next that will keep the granularity.. :-)

**Pull request closed without merging**